cmd/fpga_plugin/README.md

```diff
@@ -1,120 +1,269 @@
+        * [Verify plugin registration](#verify-plugin-registration)
+    * [Deploy by hand](#deploy-by-hand)
+        * [Build FPGA device plugin](#build-fpga-device-plugin)
+        * [Run FPGA device plugin in af mode](#run-fpga-device-plugin-in-af-mode)
+        * [Run FPGA device plugin in region mode](#run-fpga-device-plugin-in-region-mode)
+* [Next steps](#next-steps)
+
+# Introduction
+
+This FPGA device plugin is part of a collection of Kubernetes components found within this
+repository that enable integration of Intel FPGA hardware into Kubernetes.
+
+The following hardware platforms are supported:
+
+- Intel Arria 10
+- Intel Stratix 10
+
+The components support the [Open Programmable Acceleration Engine (OPAE)](https://opae.github.io/latest/index.html)
+and OpenCL application interfaces.
+
+The components together implement the following features:
+
+- discovery of pre-programmed accelerator functions
+- orchestration of FPGA programming
+- access control for FPGA hardware
+
+# Component overview
+
+The following components are part of this repository, and work together to support Intel FPGAs under
+Kubernetes:
+
+-   [FPGA device plugin](README.md) (this component)
+
+    A Kubernetes [device plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)
+    that discovers available FPGA resources on a node and advertises them to the Kubernetes control plane
+    via the node kubelet.
 
--   [FPGA device plugin](README.md) (this module)
 -   [FPGA admission controller webhook](../fpga_admissionwebhook/README.md)
+
+    A Kubernetes [admission controller webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/)
+    which can be used to dynamically convert logical resource names in pod specifications into actual FPGA
+    resource names, as advertised by the device plugin.
+
+    The webhook can also set environment variables to instruct the FPGA prestart hook to program the FPGA
+    before launching the container.
+
 -   [FPGA prestart CRI-O hook](../fpga_crihook/README.md)
 
+    A [CRI-O](https://github.com/cri-o/cri-o) prestart hook that, upon instruction from the FPGA admission
+    controller, allocates and programs the FPGA before the container is launched.
 
-### Get source code:
-```
-    $ mkdir -p $GOPATH/src/github.com/intel/
-    $ cd $GOPATH/src/github.com/intel/
-    $ git clone https://github.com/intel/intel-device-plugins-for-kubernetes.git
-```
+The repository also contains an [FPGA helper tool](../fpga_tool/README.md) that may be useful during
+development, initial deployment and debugging.
 
-### Build FPGA device plugin:
-```
-    $ cd $GOPATH/src/github.com/intel/intel-device-plugins-for-kubernetes
-    $ make fpga_plugin
-```
+# Installation
+
+The below sections cover how to obtain, build and install this component.
+
+Components can generally be installed either using DaemonSets or running them
+'by hand' on each node.
+
```

Comment:
I'm not sure where to put this better, but it should be explained that plugin daemonset contains fpga plugin container  and initcontainer. initcontainer installs prestart hook and fpgatool into /opt/intel/fpga-sw and puts hook configuration file intel-fpga-crihook.json into /etc/containers/oci/hooks.d/ on every node.

Reply:
I mention in the DaemonSet section that the DaemonSet deploys the initcontainer as well as the plugin. I'd suggest adding the details about the installation of the fpgatool, hook and config JSON file to the hook document.

Of course, what I'd really like is a nice top level diagram (or two) showing how the components install and interact in a node - but, unless we happen to have one already that I can use/copy, I don't think that will happen in this PR.

Reply:
> . I'd suggest adding the details about the installation of the fpgatool, hook and config JSON file to the hook document.

Makes sense to me.

Reply:
@grahamwhaley Would something like this go? I can make another one for preprogrammed mode if needed.

![FPGA-orchestrated](https://user-images.githubusercontent.com/13677519/72998766-0bb00680-3e07-11ea-8d5b-df14f9cd2a48.png)

Reply:
@bart0sh - yes, I think that diagram will help! I can take that and add it in at the top of the fpga plugin docs as part of the overview/introduction. A matching preprogrammed one would also of course be nice ;-)
I don't suppose the 'source' of that diagram is in some nice textual ideally FOSS form is it (dia, inkscape, UML text etc.?) :-) It would be great to have a 'source' of the diagrams checked in alongside the PNGs...

Reply:
If you don't have the source, I might try to redraw it in one of the aforementioned tools so we do have the source... np.

Reply:
I used [draw.io](https://www.draw.io) to draw it. Here is the latest version in PNG and XML format. It should be possible to import it from XML into draw.io I think. I can also share it in their own format if needed. They use my google drive to store it.

![FPGA-orchestrated](https://user-images.githubusercontent.com/13677519/73059450-71030680-3e9e-11ea-8c45-c74c36324f88.png)

Reply:
github was not able to attach XML here. Posting it as is:
```<?xml version="1.0" encoding="UTF-8"?>
<mxfile modified="2020-01-24T09:47:46.534Z" host="www.draw.io" agent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko/20100101 Firefox/71.0" etag="Jtp2FVY77_LjvggZoJqp" version="12.5.6" type="google">
  <diagram id="f106602c-feb2-e66a-4537-3a34d633f6aa" name="Page-1">
    <mxGraphModel dx="1981" dy="1122" grid="1" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" page="1" pageScale="1" pageWidth="1169" pageHeight="827" background="#ffffff" math="0" shadow="0">
      <root>
        <mxCell id="0"/>
        <mxCell id="1" parent="0"/>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-790" value="Node" style="rounded=1;absoluteArcSize=1;arcSize=20;html=1;strokeColor=none;gradientColor=none;shadow=0;dashed=0;strokeColor=none;fontSize=12;fontColor=#9E9E9E;align=left;verticalAlign=top;spacing=10;spacingTop=-4;fillColor=#E1F5FE;" parent="1" vertex="1">
          <mxGeometry x="221" y="166" width="439" height="494" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-859" value="" style="strokeColor=#dddddd;fillColor=#ffffff;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontSize=10;fontColor=#9E9E9E;align=center;html=1;" parent="1" vertex="1">
          <mxGeometry x="232" y="295" width="199" height="241" as="geometry"/>
        </mxCell>
        <mxCell id="753" value="Kubernetes" style="rounded=1;absoluteArcSize=1;arcSize=2;html=1;strokeColor=none;gradientColor=none;shadow=0;dashed=0;strokeColor=none;fontSize=12;fontColor=#9E9E9E;align=left;verticalAlign=top;spacing=10;spacingTop=-4;fillColor=#E1F5FE;" parent="1" vertex="1">
          <mxGeometry x="701" y="166" width="240" height="260" as="geometry"/>
        </mxCell>
        <mxCell id="779" value="" style="fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=1;shadow=0;gradientColor=none;fontSize=10;fontColor=#9E9E9E;align=center;html=1;dashed=1;dashPattern=1 4;" parent="1" vertex="1">
          <mxGeometry x="240.5" y="690" width="680.5" height="80" as="geometry"/>
        </mxCell>
        <mxCell id="780" value="4 Create Pod" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="200" height="30" relative="1" as="geometry">
            <mxPoint x="254.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="781" value="&lt;div&gt;5 Mutate Pod&lt;br&gt;&lt;/div&gt;&lt;div&gt;&amp;nbsp;&amp;nbsp; - map names to IDs&lt;br&gt;&lt;/div&gt;&lt;div&gt;&amp;nbsp;&amp;nbsp; - set environment for CRI-O hook&lt;br&gt;&lt;/div&gt;" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="220" height="40" relative="1" as="geometry">
            <mxPoint x="254.5" y="19" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="782" value="2 Communicate through Device plugin API" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="260" height="30" relative="1" as="geometry">
            <mxPoint y="20" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="783" value="3 Configure CRI-O to run prestart hook" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="240" height="30" relative="1" as="geometry">
            <mxPoint y="40" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="784" value="6 Program devices in prestart hook" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="230" height="20" relative="1" as="geometry">
            <mxPoint x="468.5" y="-3" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-925" value="7 Run user container" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="230" height="30" relative="1" as="geometry">
            <mxPoint x="468.5" y="20" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-926" value="8 Use programmed devices" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="230" height="30" relative="1" as="geometry">
            <mxPoint x="469.5" y="41" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-950" value="1 Discover devices in SysFS" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="240" height="30" relative="1" as="geometry">
            <mxPoint as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-956" value="FPGA Device Plugin Daemonset" style="strokeColor=none;fillColor=none;fontColor=#757575;align=left;html=1;fontStyle=0;spacingLeft=5;fontSize=11;verticalAlign=top;whiteSpace=wrap;spacingRight=5;" parent="779" vertex="1">
          <mxGeometry width="240" height="30" relative="1" as="geometry">
            <mxPoint y="-180" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-952" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.25;exitY=1;exitDx=0;exitDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;entryX=0.5;entryY=0;entryDx=0;entryDy=0;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-795" target="KYx4Pj8fMT2gdDbSYMlN-812" edge="1">
          <mxGeometry relative="1" as="geometry">
            <mxPoint x="777" y="450" as="targetPoint"/>
            <Array as="points">
              <mxPoint x="776" y="394"/>
            </Array>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-795" value="" style="strokeColor=#dddddd;fillColor=#ffffff;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontSize=10;fontColor=#9E9E9E;align=center;html=1;" parent="1" vertex="1">
          <mxGeometry x="732" y="326" width="178" height="68" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-796" value="&lt;font color=&quot;#000000&quot;&gt;API Server&lt;/font&gt;" style="dashed=0;connectable=0;html=1;fillColor=#5184F3;strokeColor=none;shape=mxgraph.gcp2.hexIcon;prIcon=apigee_api_platform;part=1;labelPosition=right;verticalLabelPosition=middle;align=left;verticalAlign=middle;spacingLeft=5;fontColor=#999999;fontSize=12;" parent="KYx4Pj8fMT2gdDbSYMlN-795" vertex="1">
          <mxGeometry y="0.5" width="44" height="39" relative="1" as="geometry">
            <mxPoint x="5" y="-19.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-819" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.5;exitY=1;exitDx=0;exitDy=0;strokeColor=#4284F3;strokeWidth=2;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-810" target="KYx4Pj8fMT2gdDbSYMlN-795" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-870" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0;exitY=0.5;exitDx=0;exitDy=0;entryX=1;entryY=0.5;entryDx=0;entryDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-810" target="KYx4Pj8fMT2gdDbSYMlN-857" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-810" value="" style="strokeColor=#dddddd;fillColor=#ffffff;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontSize=10;fontColor=#9E9E9E;align=center;html=1;" parent="1" vertex="1">
          <mxGeometry x="732" y="206" width="178" height="68" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-811" value="&lt;font color=&quot;#000000&quot;&gt;Scheduler&lt;br&gt;&lt;/font&gt;" style="dashed=0;connectable=0;html=1;fillColor=#5184F3;strokeColor=none;shape=mxgraph.gcp2.hexIcon;prIcon=apigee_api_platform;part=1;labelPosition=right;verticalLabelPosition=middle;align=left;verticalAlign=middle;spacingLeft=5;fontColor=#999999;fontSize=12;" parent="KYx4Pj8fMT2gdDbSYMlN-810" vertex="1">
          <mxGeometry y="0.5" width="44" height="39" relative="1" as="geometry">
            <mxPoint x="5" y="-19.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-856" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.75;exitY=0;exitDx=0;exitDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" edge="1">
          <mxGeometry relative="1" as="geometry">
            <mxPoint x="800" y="456" as="sourcePoint"/>
            <mxPoint x="800" y="394" as="targetPoint"/>
            <Array as="points">
              <mxPoint x="800" y="394"/>
            </Array>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-812" value="" style="strokeColor=#dddddd;fillColor=#ffffff;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontSize=10;fontColor=#9E9E9E;align=center;html=1;" parent="1" vertex="1">
          <mxGeometry x="687" y="456" width="178" height="68" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-813" value="&lt;div&gt;&lt;font color=&quot;#000000&quot;&gt;FPGA Admission&lt;/font&gt;&lt;/div&gt;&lt;div&gt;&lt;font color=&quot;#000000&quot;&gt;Controller&lt;/font&gt;&lt;/div&gt;&lt;div&gt;&lt;font color=&quot;#000000&quot;&gt;Webhook&lt;br&gt;&lt;/font&gt;&lt;/div&gt;" style="dashed=0;connectable=0;html=1;fillColor=#5184F3;strokeColor=none;shape=mxgraph.gcp2.hexIcon;prIcon=apigee_api_platform;part=1;labelPosition=right;verticalLabelPosition=middle;align=left;verticalAlign=middle;spacingLeft=5;fontColor=#999999;fontSize=12;" parent="KYx4Pj8fMT2gdDbSYMlN-812" vertex="1">
          <mxGeometry y="0.5" width="44" height="39" relative="1" as="geometry">
            <mxPoint x="5" y="-19.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-887" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0;exitY=0.5;exitDx=0;exitDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-857" target="KYx4Pj8fMT2gdDbSYMlN-881" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-897" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.5;exitY=1;exitDx=0;exitDy=0;entryX=0.5;entryY=0;entryDx=0;entryDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-857" target="KYx4Pj8fMT2gdDbSYMlN-892" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-857" value="" style="strokeColor=#dddddd;fillColor=#ffffff;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontSize=10;fontColor=#9E9E9E;align=center;html=1;" parent="1" vertex="1">
          <mxGeometry x="461" y="206" width="178" height="68" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-858" value="&lt;font color=&quot;#000000&quot;&gt;Kubelet&lt;br&gt;&lt;/font&gt;" style="dashed=0;connectable=0;html=1;fillColor=#5184F3;strokeColor=none;shape=mxgraph.gcp2.hexIcon;prIcon=apigee_api_platform;part=1;labelPosition=right;verticalLabelPosition=middle;align=left;verticalAlign=middle;spacingLeft=5;fontColor=#999999;fontSize=12;" parent="KYx4Pj8fMT2gdDbSYMlN-857" vertex="1">
          <mxGeometry y="0.5" width="44" height="39" relative="1" as="geometry">
            <mxPoint x="5" y="-19.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-949" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.192;exitY=0;exitDx=0;exitDy=0;entryX=0.5;entryY=1;entryDx=0;entryDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;exitPerimeter=0;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-881" target="KYx4Pj8fMT2gdDbSYMlN-942" edge="1">
          <mxGeometry relative="1" as="geometry">
            <Array as="points">
              <mxPoint x="276" y="310"/>
              <mxPoint x="276" y="310"/>
            </Array>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-881" value="" style="strokeColor=#dddddd;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontFamily=Helvetica;" parent="1" vertex="1">
          <mxGeometry x="241.5" y="340" width="180" height="70" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-882" value="&lt;font color=&quot;#000000&quot;&gt;Plugin&lt;/font&gt;" style="dashed=0;connectable=0;html=1;fillColor=#5184F3;strokeColor=none;shape=mxgraph.gcp2.hexIcon;prIcon=developer_portal;part=1;labelPosition=right;verticalLabelPosition=middle;align=left;verticalAlign=middle;spacingLeft=5;fontColor=#999999;fontSize=12;" parent="KYx4Pj8fMT2gdDbSYMlN-881" vertex="1">
          <mxGeometry y="0.5" width="44" height="39" relative="1" as="geometry">
            <mxPoint x="5" y="-19.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-941" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=1;exitY=0.5;exitDx=0;exitDy=0;entryX=0.25;entryY=1;entryDx=0;entryDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-883" target="KYx4Pj8fMT2gdDbSYMlN-892" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-883" value="" style="strokeColor=#dddddd;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontFamily=Helvetica;" parent="1" vertex="1">
          <mxGeometry x="241.5" y="426" width="180" height="70" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-884" value="&lt;font color=&quot;#000000&quot;&gt;InitContainer&lt;/font&gt;" style="dashed=0;connectable=0;html=1;fillColor=#5184F3;strokeColor=none;shape=mxgraph.gcp2.hexIcon;prIcon=developer_portal;part=1;labelPosition=right;verticalLabelPosition=middle;align=left;verticalAlign=middle;spacingLeft=5;fontColor=#999999;fontSize=12;" parent="KYx4Pj8fMT2gdDbSYMlN-883" vertex="1">
          <mxGeometry y="0.5" width="44" height="39" relative="1" as="geometry">
            <mxPoint x="5" y="-19.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-920" value="" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-886" target="KYx4Pj8fMT2gdDbSYMlN-857" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-886" value="2" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="411" y="230" width="20" height="20" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-788" value="" style="shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;labelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=bottom;spacingLeft=0;fontSize=12;whiteSpace=wrap;spacingBottom=2;" parent="1" vertex="1">
          <mxGeometry x="811" y="566" width="100" height="80" as="geometry"/>
        </mxCell>
        <mxCell id="756" value="Users" style="rounded=1;absoluteArcSize=1;arcSize=2;html=1;shadow=0;dashed=0;strokeColor=none;fontSize=12;align=left;verticalAlign=top;spacing=10;spacingTop=-4;" parent="KYx4Pj8fMT2gdDbSYMlN-788" vertex="1">
          <mxGeometry x="-8.736414170745848" y="-3.591631912657243" width="108.73641417074585" height="83.59163191265725" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-789" value="" style="dashed=0;connectable=0;html=1;fillColor=#757575;strokeColor=none;shape=mxgraph.gcp2.users;part=1;" parent="KYx4Pj8fMT2gdDbSYMlN-788" vertex="1">
          <mxGeometry x="0.5" width="50" height="31.5" relative="1" as="geometry">
            <mxPoint x="-25" y="19.25" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-936" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.5;exitY=1;exitDx=0;exitDy=0;entryX=1;entryY=0.25;entryDx=0;entryDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-892" target="KYx4Pj8fMT2gdDbSYMlN-931" edge="1">
          <mxGeometry relative="1" as="geometry">
            <Array as="points">
              <mxPoint x="550" y="506"/>
              <mxPoint x="451" y="506"/>
              <mxPoint x="451" y="574"/>
            </Array>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-945" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.75;exitY=1;exitDx=0;exitDy=0;entryX=0.75;entryY=0;entryDx=0;entryDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-892" target="KYx4Pj8fMT2gdDbSYMlN-898" edge="1">
          <mxGeometry relative="1" as="geometry">
            <Array as="points">
              <mxPoint x="595" y="520"/>
              <mxPoint x="560" y="520"/>
            </Array>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-892" value="" style="strokeColor=#dddddd;fillColor=#ffffff;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontSize=10;fontColor=#9E9E9E;align=center;html=1;" parent="1" vertex="1">
          <mxGeometry x="461" y="333" width="178" height="68" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-893" value="&lt;font color=&quot;#000000&quot;&gt;CRI-O&lt;br&gt;&lt;/font&gt;" style="dashed=0;connectable=0;html=1;fillColor=#5184F3;strokeColor=none;shape=mxgraph.gcp2.hexIcon;prIcon=apigee_api_platform;part=1;labelPosition=right;verticalLabelPosition=middle;align=left;verticalAlign=middle;spacingLeft=5;fontColor=#999999;fontSize=12;" parent="KYx4Pj8fMT2gdDbSYMlN-892" vertex="1">
          <mxGeometry y="0.5" width="44" height="39" relative="1" as="geometry">
            <mxPoint x="5" y="-19.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-895" value="6" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="539" y="291" width="20" height="20" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-900" value="&#xa;&#xa;&#xa;&#xa;&#xa;FPGA &#xa;device" style="strokeColor=#dddddd;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;labelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=bottom;spacingLeft=0;fontColor=#999999;fontSize=12;whiteSpace=wrap;spacingBottom=2;spacing=1;" parent="1" vertex="1">
          <mxGeometry x="539" y="556" width="72" height="90" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-901" value="" style="dashed=0;connectable=0;html=1;fillColor=#757575;strokeColor=none;shape=mxgraph.gcp2.circuit_board;part=1;" parent="KYx4Pj8fMT2gdDbSYMlN-900" vertex="1">
          <mxGeometry x="0.5" width="50" height="45" relative="1" as="geometry">
            <mxPoint x="-25" y="12.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-898" value="&#xa;&#xa;&#xa;&#xa;&#xa;FPGA &#xa;device" style="strokeColor=#dddddd;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;labelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=bottom;spacingLeft=0;fontColor=#999999;fontSize=12;whiteSpace=wrap;spacingBottom=2;spacing=1;" parent="1" vertex="1">
          <mxGeometry x="507" y="546" width="70" height="90" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-899" value="" style="dashed=0;connectable=0;html=1;fillColor=#757575;strokeColor=none;shape=mxgraph.gcp2.circuit_board;part=1;" parent="KYx4Pj8fMT2gdDbSYMlN-898" vertex="1">
          <mxGeometry x="0.5" width="50" height="45" relative="1" as="geometry">
            <mxPoint x="-25" y="12.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="775" value="6" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="586" y="425" width="20" height="17" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-924" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.25;exitY=1;exitDx=0;exitDy=0;entryX=0.75;entryY=1;entryDx=0;entryDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-812" target="KYx4Pj8fMT2gdDbSYMlN-812" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="776" value="5" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="768" y="534" width="20" height="20" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-927" value="7" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="540" y="425" width="20" height="20" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-935" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=1;exitY=0.5;exitDx=0;exitDy=0;entryX=0;entryY=0.5;entryDx=0;entryDy=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="KYx4Pj8fMT2gdDbSYMlN-931" target="KYx4Pj8fMT2gdDbSYMlN-898" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-931" value="" style="strokeColor=#dddddd;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;fontFamily=Helvetica;" parent="1" vertex="1">
          <mxGeometry x="232" y="556" width="180" height="70" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-932" value="&lt;font color=&quot;#000000&quot;&gt;User container&lt;/font&gt;" style="dashed=0;connectable=0;html=1;fillColor=#5184F3;strokeColor=none;shape=mxgraph.gcp2.hexIcon;prIcon=developer_portal;part=1;labelPosition=right;verticalLabelPosition=middle;align=left;verticalAlign=middle;spacingLeft=5;fontColor=#999999;fontSize=12;" parent="KYx4Pj8fMT2gdDbSYMlN-931" vertex="1">
          <mxGeometry y="0.5" width="44" height="39" relative="1" as="geometry">
            <mxPoint x="5" y="-19.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-934" value="8" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="431" y="581" width="20" height="20" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-906" value="3" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="494" y="426" width="20" height="20" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-942" value="SysFS" style="strokeColor=#dddddd;shadow=1;strokeWidth=1;rounded=1;absoluteArcSize=1;arcSize=2;labelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=bottom;spacingLeft=0;fontColor=#999999;fontSize=12;whiteSpace=wrap;spacingBottom=2;dashed=1;dashPattern=1 4;" parent="1" vertex="1">
          <mxGeometry x="241.5" y="200" width="68" height="80" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-943" value="" style="dashed=0;connectable=0;html=1;fillColor=#757575;strokeColor=none;shape=mxgraph.gcp2.stream;part=1;" parent="KYx4Pj8fMT2gdDbSYMlN-942" vertex="1">
          <mxGeometry x="0.5" width="50" height="41" relative="1" as="geometry">
            <mxPoint x="-25" y="14.5" as="offset"/>
          </mxGeometry>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-947" value="1" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="265.5" y="309" width="20" height="17" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-955" style="edgeStyle=orthogonalEdgeStyle;rounded=0;orthogonalLoop=1;jettySize=auto;html=1;exitX=0.75;exitY=0;exitDx=0;exitDy=0;entryX=0.854;entryY=1.059;entryDx=0;entryDy=0;entryPerimeter=0;strokeColor=#4284F3;strokeWidth=2;fontFamily=Helvetica;" parent="1" source="756" target="KYx4Pj8fMT2gdDbSYMlN-795" edge="1">
          <mxGeometry relative="1" as="geometry"/>
        </mxCell>
        <mxCell id="KYx4Pj8fMT2gdDbSYMlN-843" value="4" style="shape=ellipse;fillColor=#ffffff;strokeColor=#BDBDBD;strokeWidth=2;shadow=0;gradientColor=none;fontColor=#757575;align=center;html=1;fontStyle=1;spacingTop=-1;" parent="1" vertex="1">
          <mxGeometry x="875" y="480" width="20" height="20" as="geometry"/>
        </mxCell>
      </root>
    </mxGraphModel>
  </diagram>
</mxfile>```

Reply:
@grahamwhaley Shared 2 diagrams (for preprogrammed and orchestration programmed modes) with you on draw.io. You should be able to edit them.

Reply:
I see the diagrams, thanks. I dug into draw.io a little, to try and ensure we were not pinning ourselves to some tooling that might 'go away' in the future. I'm pretty happy with the state of source code open-ness with draw.io, so suggest we store both the `.png` and `.drawio` files in the repo. I'll add those two to this doc...